### PR TITLE
tapgarden: use lru.Cache for various new group caches

### PR DESCRIPTION
**Pull request by Roasbeef**
This also fixes a bug in `GenRawGroupAnchorVerifier` where the `groupAnchor` value wasn't being properly re assigned.

### CODE REVIEW

File: tapgarden/caretaker.go [hunk 1]
```diff
@@ -1381,12 +1400,14 @@ func GenRawGroupAnchorVerifier(ctx context.Context) func(*asset.Genesis,
 				return ErrGenesisNotGroupAnchor
 			}
 
-			groupAnchors[assetGroupKey] = gen
```
Comment (Roasbeef) [inline tapgarden/caretaker.go:1384]: So here `groupAnchor` wasn't re assigned.

File: tapgarden/caretaker.go [hunk 2]
```diff
@@ -1263,81 +1265,103 @@ func GenHeaderVerifier(ctx context.Context,
...
+
+// emptyVal is a simple type def around struct{} to use as a dummy value in in
+// the cache.
+type emptyVal struct{}
+
+// singleCacheValue is a dummy value that can be used to add an element to the
+// cache. This should be used when the cache just needs to worry aobut the
+// total number of elements, and not also the size (in bytes) of the elements.
+type singleCacheValue[T any] struct {
+	val T
```
Comment (Roasbeef) [inline tapgarden/caretaker.go:1279]: You can't embed type param values, so had to use this here.

File: tapgarden/caretaker.go [hunk 3]
```diff
@@ -1263,81 +1265,103 @@ func GenHeaderVerifier(ctx context.Context,
...
+// newSingleValue creates a new single cache value.
+func newSingleValue[T any](v T) singleCacheValue[T] {
+	return singleCacheValue[T]{
+		val: v,
+	}
+}
+
+// emptyCacheVal is a type def for an empty cache value. In this case the cache
+// is used more as a set.
+type emptyCacheVal = singleCacheValue[emptyVal]
```
Comment (Roasbeef) [inline tapgarden/caretaker.go:1296]: Type alias used here to it inherits all the methods.

File: tapgarden/caretaker.go (hunk 1 above)
Comment (Roasbeef) [reply tapgarden/caretaker.go:1384]: Ah returns nil anyway, so np.

File: tapgarden/caretaker.go [hunk 4]
```diff
@@ -1263,81 +1265,103 @@ func GenHeaderVerifier(ctx context.Context,
...
 
 	// Cache anchors for groups that were previously fetched.
-	groupAnchors := make(map[asset.SerializedKey]*asset.Genesis)
-
-	// verifierLock is a mutex that must be held when using the group
-	// anchor verifier.
-	var verifierLock sync.Mutex
+	groupAnchors := lru.NewCache[
+		asset.SerializedKey, singleCacheValue[*asset.Genesis]](
+		assetGroupCacheSize,
```
Comment (jharveyb) [inline tapgarden/caretaker.go:1345]: IIUC with a 10k element cache  this would be a few MB max?

**Review by jharveyb — APPROVED** [review]
LGTM  🚀 

Will be useful pattern for safely adding caching to some other callbacks as well.

**Pull request merged**